Sources/CodeGen/CXX/CXXTranspiler.swift

```diff
@@ -603,4 +600,49 @@ public struct CXXTranspiler {
 
   }
 
+  // MARK: entry point function body
+
+  func entryPointBody(module source: ModuleDecl.Typed) -> CXXScopedBlock? {
+    let possibleBodies = source.topLevelDecls.compactMap({
+      entryPointBody(topLevel: $0, in: source)
+    })
+    return possibleBodies.isEmpty ? nil : possibleBodies.first
+  }
+  func entryPointBody(topLevel source: AnyDeclID.TypedNode, in module: ModuleDecl.Typed)
+    -> CXXScopedBlock?
+  {
+    let function = FunctionDecl.Typed(source)
+    if function != nil && function!.identifier?.value == "main" {
+      return entryPointBody(mainFunction: FunctionDecl.Typed(source)!, in: module)
+    } else {
+      return nil
+    }
+  }
```

Comment:
This pair of function seems overly complex. I guess we're trying to avoid raw loops, but in this case it's impeding readability. Also, I don't think you need any loop in the first place, IIUC:
```suggestion
  func entryPointBody(module source: ModuleDecl.Typed) -> CXXScopedBlock? {
    if let f = decl.topLevelDecls.first(where: { (d) in
      FunctionDecl.Typed(d)?.identifier?.value == "main"
    }) {
      return entryPointBody(mainFunction: FunctionDecl.Typed(f)!, in: module)
    } else {
      return nil
    }
  }
```
It's a bit unfortunate that we have to convert the ID of the main function to `FunctionDecl.Typed` twice. Perhaps we could extend `Sequence` with a method `first(transformedBy:)` to make that simpler.

```swift
extension Sequence {

  public func first<T>(transformedBy transform: (Element) throws -> T?) rethrows -> T? {
    for x in self {
      if let y = try transform(x) { return y }
    }
    return nil
  }

}
```

Reply:
This was the abstraction I was looking for. Thank you for pointing this out!